Buildings/Controls/OBC/CDL/Logical/VariableDutyCycle.mo

```diff
@@ -0,0 +1,270 @@
+within Buildings.Controls.OBC.CDL.Logical;
+block VariableDutyCycle
+  "Produce output that cycles on and off according to the specified pulse period and width input"
+
+  parameter Real samplePeriod(
+    final quantity="Time",
+    final unit="s",
+    min=1E-3)
+    "Sample period of component";
+  parameter Real period(
+    final quantity="Time",
+    final unit="s",
+    final min=Constants.small)
+    "Time for one period";
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput uCycOn(
+    final min=0,
+    final max=1,
+    final unit="1")
+    "Ratio of the cycle time that the output should be on"
+    annotation (Placement(transformation(extent={{-140,-20},{-100,20}}),
+        iconTransformation(extent={{-140,-20},{-100,20}})));
+  Buildings.Controls.OBC.CDL.Interfaces.BooleanOutput y
+    "Connector of Boolean output signal"
+    annotation (Placement(transformation(extent={{100,-60},{140,-20}}),
+        iconTransformation(extent={{100,-20},{140,20}})));
+
+  Buildings.Controls.OBC.CDL.Discrete.Sampler sam(
+    final samplePeriod=samplePeriod)
+    "Sample the input and use it to check the input value change"
+    annotation (Placement(transformation(extent={{-80,70},{-60,90}})));
+  Buildings.Controls.OBC.CDL.Continuous.Subtract sub
+    "Check the difference before and after sampling"
+    annotation (Placement(transformation(extent={{-40,50},{-20,70}})));
+  Buildings.Controls.OBC.CDL.Continuous.Abs abs1
+    "Find the absolute value change"
+    annotation (Placement(transformation(extent={{-10,50},{10,70}})));
+  Buildings.Controls.OBC.CDL.Continuous.GreaterThreshold greThr
+    "Check if there is input value change"
+    annotation (Placement(transformation(extent={{20,50},{40,70}})));
+  Buildings.Controls.OBC.CDL.Continuous.Switch swi
+    "If there is input value change, use the new value"
+    annotation (Placement(transformation(extent={{60,20},{80,40}})));
+  Buildings.Controls.OBC.CDL.Continuous.GreaterThreshold greThr1
```

Comment:
This needs a hysteresis, else if uCycOn is the output of a PI controller, noise will trigger it.